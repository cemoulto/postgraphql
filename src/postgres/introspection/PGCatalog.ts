import PGObject from './object/PGObject'
import PGNamespaceObject from './object/PGNamespaceObject'
import PGClassObject from './object/PGClassObject'
import PGAttributeObject from './object/PGAttributeObject'
import PGTypeObject from './object/PGTypeObject'

/**
 * A utility class for interacting with the `PGObject`s returned from the
 * introspection query.
 */
class PGCatalog {
  private _namespaces = new Map<string, PGNamespaceObject>()
  private _classes = new Map<string, PGClassObject>()
  private _attributes = new Map<string, PGAttributeObject>()
  private _types = new Map<string, PGTypeObject>()

  constructor (objects: Array<PGObject>) {
    // Build an in-memory index of all our objects for ease of use:
    for (const object of objects) {
      switch (object.kind) {
        case 'namespace':
          this._namespaces.set(object.id, object)
          break
        case 'class':
          this._classes.set(object.id, object)
          break
        case 'attribute':
          this._attributes.set(`${object.classId}-${object.num}`, object)
          break
        case 'type':
          this._types.set(object.id, object)
          break
        default:
          throw new Error(`Object of kind '${object['kind']}' is not allowed.`)
      }
    }
  }

  /**
   * Gets all of the namespace objects.
   */
  public getNamespaces (): Array<PGNamespaceObject> {
    return Array.from(this._namespaces.values())
  }

  /**
   * Gets a single namespace object of the provided id.
   */
  public getNamespace (id: string): PGNamespaceObject | undefined {
    return this._namespaces.get(id)
  }

  /**
   * Gets a single namespace object by the provided id, and if no namespace
   * object exists an error is thrown instead of returning `undefined`.
   */
  public assertGetNamespace (id: string): PGNamespaceObject {
    const namespace = this.getNamespace(id)

    if (!namespace)
      throw new Error(`No namespace was found with id ${id}`)

    return namespace
  }

  /**
   * Gets all of the class objects.
   */
  public getClasses (): Array<PGClassObject> {
    return Array.from(this._classes.values())
  }

  /**
   * Gets a single class object of the provided id.
   */
  public getClass (id: string): PGClassObject | undefined {
    return this._classes.get(id)
  }

  /**
   * Gets a single class object by the provided id, and if no class object
   * exists an error is thrown instead of returning `undefined`.
   */
  public assertGetClass (id: string): PGClass {
    const clazz = this.getClass(id)

    if (!clazz)
      throw new Error(`No class was found with id ${id}`)

    return clazz
  }

  /**
   * Gets all of the attribute objects.
   */
  public getAttributes (): Array<PGAttributeObject> {
    return Array.from(this._attributes.values())
  }

  /**
   * Gets a single attribute object by the provided class id and number.
   */
  public getAttribute (classId: string, num: number): PGAttributeObject | undefined {
    return this._attributes.get(`${classId}-${num}`)
  }

  /**
   * Gets all of the attributes for a single class.
   */
  public getClassAttributes (classId: string): Array<PGAttributeObject> {
    return Array.from(this._attributes.values()).filter(attribute => attribute.classId === classId)
  }

  /**
   * Gets all of the type objects.
   */
  public getTypes (): Array<PGTypeObject> {
    return Array.from(this._types.values())
  }

  /**
   * Gets a single type object by the provided id.
   */
  public getType (id: string): PGTypeObject | undefined {
    return this._types.get(id)
  }

  /**
   * Determines if our instance has this *exact* `PGType` instance.
   */
  public hasType (type: PGTypeObject): boolean {
    return this._types.get(type.id) === type
  }

  /**
   * Gets a single type object by the provided id, and if no type object
   * exists an error is thrown instead of returning `undefined`.
   */
  public assertGetType (id: string): PGTypeObject {
    const type = this.getType(id)

    if (!type)
      throw new Error(`No type was found with id ${id}`)

    return type
  }
}

export default PGCatalog
