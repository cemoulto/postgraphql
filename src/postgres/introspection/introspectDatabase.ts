import { resolve } from 'path'
import { readFileSync } from 'fs'
import { Client } from 'pg'
import PGCatalog from './PGCatalog'

/**
 * The introspection query SQL string. We read this from it’s SQL file
 * synchronously at runtime. It’s just like requiring a file, except that file
 * is SQL.
 */
export const introspectionQuery =
  readFileSync(resolve(__dirname, '../../../queries/introspection.sql')).toString()

/**
 * Takes a PostgreSQL client and introspects it, returning an instance of
 * `PGObjects` which can then be consumed. Note that some translation is done
 * from the raw PostgreSQL catalog to the friendlier `PGObjects` interface.
 */
export default async function introspectDatabase (client: Client, schemas: string[]): Promise<PGCatalog> {
  // Run our single introspection query in the database.
  const result = await client.query({
    name: 'introspectionQuery',
    text: introspectionQuery,
    values: [schemas],
  })

  // Extract out the objects from the query.
  const objects = new PGCatalog(result.rows.map(({ object }) => object))

  return objects
}
