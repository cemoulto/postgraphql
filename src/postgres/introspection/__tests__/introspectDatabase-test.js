import { resolve } from 'path'
import { readFileSync } from 'fs'
import { Client } from 'pg'
import PGCatalog from '../PGCatalog'
import introspectDatabase from '../introspectDatabase'

const testSchema = readFileSync(resolve(__dirname, 'fixtures/kitchen-sink-schema.sql')).toString()

/**
 * Gets a local identifier that is independent of the object id assigned by
 * PostgreSQL which will be consistent across tests.
 *
 * Just a concatenation of the PostgreSQL objectâ€™s kind and name. For
 * attributes we also include the number.
 */
const getLocalId = pgObject => {
  if (pgObject.kind === 'attribute')
    return `${pgObject.kind}-${pgObject.name}-${pgObject.num}`

  return `${pgObject.kind}-${pgObject.name}`
}

/**
 * A utility for the following `format` function that creates a helper to
 * replace certain properties with their local id equivalents (see
 * `getLocalId` above).
 */
const withLocalIds = properties => pgObject => {
  const pgObjectLocalIds = {}

  for (const property of properties)
    if (pgObject[property] != null)
      pgObjectLocalIds[property] = getLocalId(pgObject)

  return Object.assign({}, pgObject, pgObjectLocalIds)
}

/**
 * A utility function to help sort arrays. Almost identical to the Lodash
 * `_.sortBy` function.
 */
const sortBy = getKey => (a, b) => {
  const aKey = getKey(a)
  const bKey = getKey(b)
  if (aKey > bKey) return 1
  if (aKey < bKey) return -1
  return 0
}

/**
 * Formats a `PGCatalog` object into a form that is easily snapshotable by
 * Jest. This gives us all the benefits of snapshot testing.
 */
const format = catalog => ({
  namespaces: Array.from(catalog._namespaces.values())
    .map(namespace => Object.assign({}, namespace, {
      id: namespace.name,
    }))
    .sort(sortBy(({ id }) => id)),

  classes: Array.from(catalog._classes.values())
    .map(klass => Object.assign({}, klass, {
      id: klass.name,
      namespaceId: catalog.getNamespace(klass.namespaceId).name,
      typeId: catalog.getType(klass.typeId).name,
    }))
    .sort(sortBy(({ id }) => id)),

  attributes: Array.from(catalog._attributes.values())
    .map(attribute => Object.assign({}, attribute, {
      classId: catalog.getClass(attribute.classId).name,
      typeId: catalog.getType(attribute.typeId).name,
    }))
    .sort(sortBy(({ classId, num }) => `${classId}-${num}`)),

  types: Array.from(catalog._types.values())
    .map(type => Object.assign({}, type, {
      id: type.name,
      namespaceId: catalog.getNamespace(type.namespaceId) ? catalog.getNamespace(type.namespaceId).name : 'external',
      classId: type.classId ? catalog.getClass(type.classId).name : null,
      baseTypeId: type.baseTypeId ? catalog.getType(type.baseTypeId) : null,
    }))
    .sort(sortBy(({ id }) => id)),
})

test('will get everything needed in an introspection', async () => {
  const client = new Client({
    database: 'postgraphql_test',
    port: 5432,
  })

  await client.connect()
  await client.query(testSchema)

  expect(format(await introspectDatabase(client, ['a', 'b', 'c']))).toMatchSnapshot()
  expect(format(await introspectDatabase(client, ['a']))).toMatchSnapshot()

  client.end()
})
