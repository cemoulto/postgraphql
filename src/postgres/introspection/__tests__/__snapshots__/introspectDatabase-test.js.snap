exports[`test will get everything needed in an introspection 1`] = `
Object {
  "attributes": Array [
    Object {
      "classId": "compound_type",
      "description": null,
      "hasDefault": false,
      "isNotNull": false,
      "kind": "attribute",
      "name": "a",
      "num": 1,
      "typeId": "int4"
    },
    Object {
      "classId": "compound_type",
      "description": null,
      "hasDefault": false,
      "isNotNull": false,
      "kind": "attribute",
      "name": "b",
      "num": 2,
      "typeId": "text"
    },
    Object {
      "classId": "compound_type",
      "description": null,
      "hasDefault": false,
      "isNotNull": false,
      "kind": "attribute",
      "name": "c",
      "num": 3,
      "typeId": "color"
    },
    Object {
      "classId": "no_update",
      "description": null,
      "hasDefault": false,
      "isNotNull": false,
      "kind": "attribute",
      "name": "?column?",
      "num": 1,
      "typeId": "int4"
    },
    Object {
      "classId": "person",
      "description": null,
      "hasDefault": true,
      "isNotNull": true,
      "kind": "attribute",
      "name": "id",
      "num": 1,
      "typeId": "int4"
    },
    Object {
      "classId": "person",
      "description": "The personâ€™s name",
      "hasDefault": false,
      "isNotNull": true,
      "kind": "attribute",
      "name": "name",
      "num": 2,
      "typeId": "varchar"
    },
    Object {
      "classId": "person",
      "description": null,
      "hasDefault": false,
      "isNotNull": false,
      "kind": "attribute",
      "name": "about",
      "num": 3,
      "typeId": "text"
    },
    Object {
      "classId": "yo",
      "description": null,
      "hasDefault": false,
      "isNotNull": false,
      "kind": "attribute",
      "name": "__id",
      "num": 1,
      "typeId": "int4"
    },
    Object {
      "classId": "yo",
      "description": null,
      "hasDefault": false,
      "isNotNull": false,
      "kind": "attribute",
      "name": "name",
      "num": 2,
      "typeId": "varchar"
    },
    Object {
      "classId": "yo",
      "description": null,
      "hasDefault": false,
      "isNotNull": false,
      "kind": "attribute",
      "name": "description",
      "num": 3,
      "typeId": "text"
    },
    Object {
      "classId": "yo",
      "description": "This is constantly 2",
      "hasDefault": false,
      "isNotNull": false,
      "kind": "attribute",
      "name": "constant",
      "num": 4,
      "typeId": "int4"
    }
  ],
  "classes": Array [
    Object {
      "description": null,
      "id": "compound_type",
      "isDeletable": false,
      "isInsertable": false,
      "isSelectable": false,
      "isUpdatable": false,
      "kind": "class",
      "name": "compound_type",
      "namespaceId": "c",
      "typeId": "compound_type"
    },
    Object {
      "description": null,
      "id": "no_update",
      "isDeletable": false,
      "isInsertable": false,
      "isSelectable": true,
      "isUpdatable": false,
      "kind": "class",
      "name": "no_update",
      "namespaceId": "a",
      "typeId": "no_update"
    },
    Object {
      "description": "Person test comment",
      "id": "person",
      "isDeletable": true,
      "isInsertable": true,
      "isSelectable": true,
      "isUpdatable": true,
      "kind": "class",
      "name": "person",
      "namespaceId": "c",
      "typeId": "person"
    },
    Object {
      "description": "YOYOYO!!",
      "id": "yo",
      "isDeletable": true,
      "isInsertable": true,
      "isSelectable": true,
      "isUpdatable": true,
      "kind": "class",
      "name": "yo",
      "namespaceId": "b",
      "typeId": "yo"
    }
  ],
  "namespaces": Array [
    Object {
      "description": "The a schema.",
      "id": "a",
      "kind": "namespace",
      "name": "a"
    },
    Object {
      "description": "qwerty",
      "id": "b",
      "kind": "namespace",
      "name": "b"
    },
    Object {
      "description": null,
      "id": "c",
      "kind": "namespace",
      "name": "c"
    }
  ],
  "types": Array [
    Object {
      "baseTypeId": null,
      "category": "E",
      "classId": null,
      "description": null,
      "enumVariants": Array [
        "red",
        "green",
        "blue"
      ],
      "id": "color",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "color",
      "namespaceId": "b",
      "type": "e"
    },
    Object {
      "baseTypeId": null,
      "category": "C",
      "classId": "compound_type",
      "description": "Awesome feature!",
      "enumVariants": null,
      "id": "compound_type",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "compound_type",
      "namespaceId": "c",
      "type": "c"
    },
    Object {
      "baseTypeId": null,
      "category": "N",
      "classId": null,
      "description": "-2 billion to 2 billion integer, 4-byte storage",
      "enumVariants": null,
      "id": "int4",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "int4",
      "namespaceId": "external",
      "type": "b"
    },
    Object {
      "baseTypeId": null,
      "category": "C",
      "classId": "no_update",
      "description": null,
      "enumVariants": null,
      "id": "no_update",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "no_update",
      "namespaceId": "a",
      "type": "c"
    },
    Object {
      "baseTypeId": null,
      "category": "C",
      "classId": "person",
      "description": null,
      "enumVariants": null,
      "id": "person",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "person",
      "namespaceId": "c",
      "type": "c"
    },
    Object {
      "baseTypeId": null,
      "category": "S",
      "classId": null,
      "description": "variable-length string, no limit specified",
      "enumVariants": null,
      "id": "text",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "text",
      "namespaceId": "external",
      "type": "b"
    },
    Object {
      "baseTypeId": null,
      "category": "D",
      "classId": null,
      "description": "date and time with time zone",
      "enumVariants": null,
      "id": "timestamptz",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "timestamptz",
      "namespaceId": "external",
      "type": "b"
    },
    Object {
      "baseTypeId": null,
      "category": "S",
      "classId": null,
      "description": "varchar(length), non-blank-padded string, variable storage length",
      "enumVariants": null,
      "id": "varchar",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "varchar",
      "namespaceId": "external",
      "type": "b"
    },
    Object {
      "baseTypeId": null,
      "category": "C",
      "classId": "yo",
      "description": null,
      "enumVariants": null,
      "id": "yo",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "yo",
      "namespaceId": "b",
      "type": "c"
    }
  ]
}
`;

exports[`test will get everything needed in an introspection 2`] = `
Object {
  "attributes": Array [
    Object {
      "classId": "no_update",
      "description": null,
      "hasDefault": false,
      "isNotNull": false,
      "kind": "attribute",
      "name": "?column?",
      "num": 1,
      "typeId": "int4"
    }
  ],
  "classes": Array [
    Object {
      "description": null,
      "id": "no_update",
      "isDeletable": false,
      "isInsertable": false,
      "isSelectable": true,
      "isUpdatable": false,
      "kind": "class",
      "name": "no_update",
      "namespaceId": "a",
      "typeId": "no_update"
    }
  ],
  "namespaces": Array [
    Object {
      "description": "The a schema.",
      "id": "a",
      "kind": "namespace",
      "name": "a"
    }
  ],
  "types": Array [
    Object {
      "baseTypeId": null,
      "category": "N",
      "classId": null,
      "description": "-2 billion to 2 billion integer, 4-byte storage",
      "enumVariants": null,
      "id": "int4",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "int4",
      "namespaceId": "external",
      "type": "b"
    },
    Object {
      "baseTypeId": null,
      "category": "C",
      "classId": "no_update",
      "description": null,
      "enumVariants": null,
      "id": "no_update",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "no_update",
      "namespaceId": "a",
      "type": "c"
    },
    Object {
      "baseTypeId": null,
      "category": "S",
      "classId": null,
      "description": "variable-length string, no limit specified",
      "enumVariants": null,
      "id": "text",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "text",
      "namespaceId": "external",
      "type": "b"
    },
    Object {
      "baseTypeId": null,
      "category": "D",
      "classId": null,
      "description": "date and time with time zone",
      "enumVariants": null,
      "id": "timestamptz",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "timestamptz",
      "namespaceId": "external",
      "type": "b"
    },
    Object {
      "baseTypeId": null,
      "category": "S",
      "classId": null,
      "description": "varchar(length), non-blank-padded string, variable storage length",
      "enumVariants": null,
      "id": "varchar",
      "isNotNull": false,
      "itemId": null,
      "kind": "type",
      "name": "varchar",
      "namespaceId": "external",
      "type": "b"
    }
  ]
}
`;
